pkg/versioning/versioning.go

```diff
@@ -0,0 +1,54 @@
+package versioning
+
+import (
+	"fmt"
+
+	"github.com/SAP/jenkins-library/pkg/maven"
+)
+
+// Artifact ...
+type Artifact interface {
+	VersioningScheme() string
+	GetVersion() (string, error)
+	SetVersion(string) error
+}
+
+// Options ...
+type Options struct {
+	ProjectSettingsFile string
+	GlobalSettingsFile  string
+	M2Path              string
+}
+
+type mvnRunner struct{}
+
+func (m *mvnRunner) Execute(options *maven.ExecuteOptions, execRunner mavenExecRunner) (string, error) {
+	return maven.Execute(options, execRunner)
+}
+func (m *mvnRunner) Evaluate(pomFile, expression string, execRunner mavenExecRunner) (string, error) {
+	return maven.Evaluate(pomFile, expression, execRunner)
+}
+
+// GetArtifact ...
+func GetArtifact(buildTool, buildDescriptorFilePath string, opts *Options, execRunner mavenExecRunner) (Artifact, error) {
+	var artifact Artifact
+	switch buildTool {
+	case "maven":
+		artifact = &Maven{
+			Runner:              &mvnRunner{},
+			ExecRunner:          execRunner,
+			PomPath:             buildDescriptorFilePath,
+			ProjectSettingsFile: opts.ProjectSettingsFile,
+			GlobalSettingsFile:  opts.GlobalSettingsFile,
+			M2Path:              opts.M2Path,
+		}
+	case "npm":
+		artifact = &Npm{
+			PackageJSONPath: buildDescriptorFilePath,
+		}
+	default:
+		return artifact, fmt.Errorf("build tool '%v' not supported", buildTool)
+	}
+
```

Comment:
no, this is just an initialized artifact `var artifact Artifact`

initialization is not exposed and currently called through `SetVersion()`, `GetVersion()`

Another option is to expose the `Init()` function via the interface and call it from here (versioning.go).
What do you think is better?

Reply:
I just missed that it was done at the start of the public methods. I think it is better like it is currently done, since it enforces a valid state for the public fields.